doc/rtd/howto/debugging.rst

```diff
@@ -0,0 +1,142 @@
+.. _how_to_debug:
+
+How to debug cloud-init
+***********************
+
+There are several cloud-init failure modes that one may need to
+debug. Debugging is specific to the scenario, but the starting points are often
+similar.
+
+I can't log in to my instance
+=============================
+
+One of the more challenging scenarios to debug is when you don't have
+shell access to your instance. You have a few options:
+
+1. Acquire console log messages and check for errors.
+
+2. Try running the same user-data locally, such as in one of the
+   :ref:`tutorials<tutorial_index>`. Use LXD or Qemu locally to get a shell or
+   logs then debug with :ref:`these steps<did_not_do_the_thing>`.
+
+3. Try copying the image to your local system, mount the filesystem locally
+   and inspect (create a chroot in the image filesystem before attempting the
+   steps in other sections).
+
+Cloud-init did not run
+======================
+
+1. Check the output of `cloud-init status --format json`
+   - what is the value of the `'extended_status'` key?
+   - what is the value of the `'boot_status_code'` key?
+
+2. Check the contents of :file:`/run/cloud-init/ds-identify.log`
+
+3. Check the status of the services
+
+.. code-block::
+
+   systemctl status cloud-init-local.service cloud-init.service\
+        cloud-config.service cloud-final.service
+
+.. _did_not_do_the_thing:
+
+Cloud-init ran, but didn't do what I want it to
+===============================================
+
+1. If you are using cloud-init's user data
+   :ref:`cloud config<user_data_formats-cloud_config>`, make sure
+   to :ref:`validate your user data cloud config<check_user_data_cloud_config>`
+
+2. Check for errors in `cloud-init status --long`
+   - what is the value of the `'errors'` key?
+   - what is the value of the `'recoverable_errors'` key?
+
+3. For more context on errors, check the logs files:
+   - :file:`/var/log/cloud-init.log`
+   - :file:`/var/log/cloud-init-output.log`
+
+   Identify errors in the logs and the lines preceeding these errors. Ask
+   yourself:
+
+   - According to the log files, what went wrong?
+   - How does the cloud-init error relate to the configuration provided
+     to this instance?
+   - What does the documentation say about the parts of the configuration that
+     relate to this error?
```

Comment:
> Do all of the errors raised via the logs already have corresponding docs to explain how to fix them, or is this something we need to address as well?

No, because we dont't attempt to foresee all of the possible ways that a user might pass an invalid configuration.

This is an attempt to get the user to think about whether their configuration actually makes sense and get them to actually go read the docs (many users don't).